src/convert/replacements.ts

```diff
@@ -0,0 +1,192 @@
+  for (const replacement of replacements) {
+    // TODO: node 16+ has String.replaceAll for non-regex scenarios
+    const regex =
+      typeof replacement.toReplace === 'string' ? new RegExp(replacement.toReplace, 'g') : replacement.toReplace;
+    const replaced = output.replace(regex, replacement.replaceWith);
+    if (replaced === output) {
+      // replacements need to be done sequentially
+      // eslint-disable-next-line no-await-in-loop
+      await Lifecycle.getInstance().emitWarning(
+        `Your sfdx-project.json specifies that ${replacement.toReplace.toString()} should be replaced, but it was not found.`
+      );
+    }
+    output = replaced;
+  }
+  return output;
+};
+
+/**
+ * Reads the project, gets replacements, removes an that aren't applicable due ot environment conditionals, and returns an instance of the ReplacementMarkingStream
+ */
+export const getReplacementMarkingStream = async (): Promise<ReplacementMarkingStream | undefined> => {
+  // remove any that don't agree with current env
+  const filteredReplacements = envFilter(await readReplacementsFromProject());
+  if (filteredReplacements.length) {
+    return new ReplacementMarkingStream(filteredReplacements);
+  }
+};
+
+/**
+ * Stream for marking replacements on a component.
+ * Returns a mutated component with a `replacements` property if any replacements are found.
+ * Throws if any replacements reference a file or env that does not exist
+ */
+class ReplacementMarkingStream extends Transform {
+  public constructor(private readonly replacementConfigs: ReplacementConfig[]) {
+    super({ objectMode: true });
+  }
+
+  public async _transform(
+    chunk: SourceComponent,
+    encoding: string,
+    callback: (err: Error, data: SourceComponent) => void
+  ): Promise<void> {
+    let err: Error;
+    // if deleting, or no configs, just pass through
+    if (!chunk.isMarkedForDelete() && this.replacementConfigs?.length) {
+      try {
+        chunk.replacements = await getReplacements(chunk, this.replacementConfigs);
+      } catch (e) {
+        if (!(e instanceof Error)) {
+          throw e;
+        }
+        err = e;
+      }
+    }
+    callback(err, chunk);
+  }
+}
+
+export const getContents = async (path: string): Promise<string> => {
+  if (!fileContentsCache.has(path)) {
+    try {
+      fileContentsCache.set(path, await readFile(path, 'utf8'));
+    } catch (e) {
+      throw new SfError(
+        `The file "${path}" specified in the "replacements" property of sfdx-project.json could not be read.`
+      );
+    }
+  }
+  return fileContentsCache.get(path);
+};
+
+/**
+ * Build the replacements property for a sourceComponent
+ */
+export const getReplacements = async (
+  cmp: SourceComponent,
+  replacementConfigs: ReplacementConfig[] = []
+): Promise<SourceComponent['replacements']> => {
+  // all possible filenames for this component
+  const filenames = [cmp.xml, ...cmp.walkContent()].filter(Boolean);
+  // eslint-disable-next-line no-console
```

Comment:
not needed?

I'm finding L136-150 really confusing, could be due to GH viewer, but a few comments could go a long way